lib/services/downloads_service.dart

```diff
@@ -0,0 +1,1415 @@
+  /// Useful for artists/genres, which may need to be shown in several libraries.
+  Future<List<DownloadStub>> getAllCollections(
+      {String? nameFilter,
+      BaseItemDtoType? baseTypeFilter,
+      BaseItemDto? relatedTo,
+      bool fullyDownloaded = false,
+      String? viewFilter,
+      String? childViewFilter,
+      bool nullableViewFilters = true}) {
+    return _isar.downloadItems
+        .where()
+        .typeEqualTo(DownloadItemType.collection)
+        .filter()
+        .optional(nameFilter != null,
+            (q) => q.nameContains(nameFilter!, caseSensitive: false))
+        .optional(baseTypeFilter != null,
+            (q) => q.baseItemTypeEqualTo(baseTypeFilter!))
+        .optional(
+            relatedTo != null,
+            (q) => q.infoFor((q) => q.info((q) => q.isarIdEqualTo(
+                DownloadStub.getHash(
+                    relatedTo!.id, DownloadItemType.collection)))))
+        .optional(fullyDownloaded,
+            (q) => q.not().stateEqualTo(DownloadItemState.notDownloaded))
+        .optional(
+            viewFilter != null,
+            (q) => q.group((q) => q.viewIdEqualTo(viewFilter).optional(
+                nullableViewFilters, (q) => q.or().viewIdEqualTo(null))))
+        .optional(
+            childViewFilter != null,
+            (q) => q.infoFor((q) => q.group((q) => q
+                .viewIdEqualTo(childViewFilter)
+                .optional(
+                    nullableViewFilters, (q) => q.or().viewIdEqualTo(null)))))
+        .findAll();
+  }
+
+  /// Get information about a downloaded song by BaseItemDto or id.
+  Future<DownloadStub?> getSongInfo({BaseItemDto? item, String? id}) {
+    assert((item == null) != (id == null));
+    return _isar.downloadItems
+        .get(DownloadStub.getHash(id ?? item!.id, DownloadItemType.song));
+  }
+
+  /// Get information about a downloaded collection by BaseItemDto or id.
+  Future<DownloadStub?> getCollectionInfo({BaseItemDto? item, String? id}) {
+    assert((item == null) != (id == null));
+    return _isar.downloadItems
+        .get(DownloadStub.getHash(id ?? item!.id, DownloadItemType.collection));
+  }
+
+  /// Get a song's DownloadItem by BaseItemDto or id.  This method performs file
+  /// verification and should only be used when the downloaded file is actually
+  /// needed, such as when building MediaItems.  Otherwise, [getSongInfo] should
+  /// be used instead.
+  Future<DownloadItem?> getSongDownload({BaseItemDto? item, String? id}) {
+    assert((item == null) != (id == null));
+    return _getDownloadByID(id ?? item!.id, DownloadItemType.song);
+  }
+
+  /// Get an image's DownloadItem by BaseItemDto or id.  This method performs file
+  /// verification and should only be used when the downloaded file is actually
+  /// needed, such as when building ImageProviders.
+  Future<DownloadItem?> getImageDownload(
+      {BaseItemDto? item, String? blurHash}) {
+    assert((item?.blurHash == null) != (blurHash == null));
```

Comment:
This is making sure exactly one of the inputs is null.  I basically didn't want to have both a getImagedownloadByBlurhash and a getImagedownloadByBaseItemDto function that each took one argument, so there's one function and you supply exactly one of the two arguments.

Reply:
But it wouldn't hurt if both were supplied, as long as they match, correct?

Reply:
Theoretically no, but there's no reason to want to do that.  Either you have the item, in which case that's easier, or you don't and have to supply the id.

Reply:
I'm just worried that the assertion might throw just because in the future someone decides to supply both, simply because they have both and it isn't immediately clear that the BaseItemDto is just used to get the blurhash. If there's no problem, there's no need to throw an assertion error.  
Would it make sense for the function to only accept the blurhash string, and shifting the need to extract the blurhash from the BaseItemDto to the method consumer?

Reply:
The idea is that if you see it always throw during testing and just choose one.  If it checked if they matched it might be fine in testing but then throw later if the calling code messes up and has mismatching items somehow.  Taking items limits the scope of things like how we do blurhash->image id fallback, so I'd rather not force all callers to use string ids.  I'm thinking I add that you give exactly argument to the function description.  If you really don't like that I can just split the function.

Reply:
No, you convinced me ^^  
But yes, please go ahead and add some docs that only only one of the arguments may be supplied!